### PR TITLE
DEV: chat streaming

### PR DESCRIPTION
**Pull request by jjaffeux**
This commit introduces the possibility to stream messages. To allow plugins to use streaming this commit also ships a `ChatSDK` library to allow to interact with few parts of discourse chat.

```ruby
ChatSDK::Message.create_with_stream(raw: "test") do |helper|
  5.times do |i|
    is_streaming = helper.stream(raw: "more #{i}")
    next if !is_streaming
    sleep 2
  end
end
```

This commit also introduces all the frontend parts:
- messages can now be marked as streaming
- when streaming their content will be updated when a new content is appended
- a special UI will be showing (a blinking indicator)
- a cancel button allows the user to stop the streaming, when cancelled `helper.stream(...)` will return `false`, and the plugin can decide exit early

### CODE REVIEW

File: plugins/chat/lib/chat_s_d_k/channel.rb
Comment (CvX) [inline plugins/chat/lib/chat_s_d_k/channel.rb:1]: re: filenames - you might want to add an entry to config/initializers/000-zeitwerk.rb 🙂

Comment (CvX) [reply plugins/chat/lib/chat_s_d_k/channel.rb:1]: yeah, to avoid `s_d_k` 😃

File: plugins/chat/app/services/chat/publisher.rb [hunk 1]
```diff
@@ -113,12 +113,12 @@ def self.publish_processed!(chat_message)
       )
     end
 
-    def self.publish_edit!(chat_channel, chat_message)
+    def self.publish_edit!(chat_channel, chat_message, streaming: false)
```
Comment (dbattersby) [inline plugins/chat/app/services/chat/publisher.rb:116]: Do we need to set streaming here, I'm wondering if updating a message would ever be streamed?

File: plugins/chat/app/services/chat/update_message.rb [hunk 2]
```diff
@@ -98,6 +100,8 @@ def save_message(message:, **)
     end
 
     def save_revision(message:, guardian:, **)
+      return false if message.streaming_before_last_save
```
Comment (dbattersby) [inline plugins/chat/app/services/chat/update_message.rb:103]: I can't find where `streaming_before_last_save` is defined, maybe this should just be ` if message.streaming`?

File: plugins/chat/app/services/chat/update_message.rb [hunk 3]
```diff
@@ -134,7 +138,8 @@ def should_create_revision(new_message, prev_message, guardian)
     def publish(message:, guardian:, contract:, **)
       edit_timestamp = context.revision&.created_at&.iso8601(6) || Time.zone.now.iso8601(6)
 
-      ::Chat::Publisher.publish_edit!(message.chat_channel, message)
+      ::Chat::Publisher.publish_edit!(message.chat_channel, message, streaming: contract.streaming)
```
Comment (dbattersby) [inline plugins/chat/app/services/chat/update_message.rb:141]: Same as before but just need to decide if updating the message requires us to set streaming as `true` at any point

File: plugins/chat/assets/javascripts/discourse/components/chat-message.gjs [hunk 4]
```diff
@@ -484,6 +485,19 @@ export default class ChatMessage extends Component {
     return this.args.context === MESSAGE_CONTEXT_THREAD;
   }
 
+  get shouldRenderStopMessageStreamingButton() {
+    return (
+      this.args.message.streaming &&
+      (this.currentUser.admin ||
+        this.args.message.user.id === this.currentUser.id)
```
Comment (dbattersby) [inline plugins/chat/assets/javascripts/discourse/components/chat-message.gjs:492]: Does this need to account for `in_reply_to`?

File: plugins/chat/lib/chat_sdk/message.rb [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+      new.create(**params, &block)
+    end
+
+    # Creates a new message with streaming enabled by default.
+    #
+    # This method is a convenience wrapper around `create` with `streaming: true` set by default.
+    # It supports all the same parameters and block usage as `create`.
+    #
+    # @see #create
+    def self.create_with_stream(**params, &block)
```
Comment (dbattersby) [inline plugins/chat/lib/chat_sdk/message.rb:38]: This is really nice 👍

File: plugins/chat/lib/chat_sdk/channel.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        guardian: guardian,
+        **params,
+        direction: "future",
+      ) do
+        on_success { result.messages }
+        on_failure do
+          p Chat::StepsInspector.new(result)
+          raise "Unexpected error"
+        end
+        on_failed_policy(:can_view_channel) { |contract| raise "Guardian can't view channel" }
```
Comment (dbattersby) [inline plugins/chat/lib/chat_sdk/channel.rb:33]: We can probably remove the contract here

File: plugins/chat/lib/chat_sdk/message.rb [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+          Chat::CreateMessage,
+          message: raw,
+          guardian: guardian,
+          chat_channel_id: channel_id,
+          in_reply_to_id: in_reply_to_id,
+          thread_id: thread_id,
+          upload_ids: upload_ids,
+          streaming: streaming,
+        ) do
+          on_model_not_found(:channel) { raise "Couldn't find channel with id: `#{channel_id}`" }
```
Comment (dbattersby) [inline plugins/chat/lib/chat_sdk/message.rb:63]: Not sure about this but should we support multiple languages on the SDK errors? Should we have a translation file for the SDK?

File: plugins/chat/lib/chat_sdk/message.rb [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+          on_model_not_found(:channel_membership) do
+            raise "User with id: `#{guardian.user.id}}` has no membership to this channel"
+          end
+          on_failed_policy(:ensure_valid_thread_for_channel) do |contract|
```
Comment (dbattersby) [inline plugins/chat/lib/chat_sdk/message.rb:67]: Might not need to pass the contract here too

File: plugins/chat/spec/lib/chat_s_d_k/channel_spec.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
```
Comment (dbattersby) [inline plugins/chat/spec/lib/chat_s_d_k/channel_spec.rb:1]: Maybe we can change the dir name here to `chat_sdk` too?

File: plugins/chat/spec/lib/chat_s_d_k/channel_spec.rb [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+      expect(messages).to eq([message_1])
+    end
+
+    context "when guardian can't see the channel" do
+      fab!(:channel_1) { Fabricate(:private_category_channel) }
+
+      it "fails" do
+        params[:guardian] = Fabricate(:user).guardian
+
+        expect { described_class.messages(**params) }.to raise_error("Guardian can't view channel")
```
Comment (dbattersby) [inline plugins/chat/spec/lib/chat_s_d_k/channel_spec.rb:31]: Maybe we can also test the policy for `target_message_exists`?

File: plugins/chat/spec/requests/chat/api/channels_messages_streaming_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+        expect(response.status).to eq(403)
+      end
+    end
+
+    context "when the user can stop" do
+      fab!(:current_user) { Fabricate(:admin) }
+      fab!(:message_1) { Fabricate(:chat_message, chat_channel: channel_1) }
+
+      it "return a 400 error" do
```
Comment (dbattersby) [inline plugins/chat/spec/requests/chat/api/channels_messages_streaming_controller_spec.rb:59]: small typo, should be `200`

**Comment by dbattersby** [comment]
Looks really good, added some small details. Looking forward to seeing this live 🚀

File: plugins/chat/app/services/chat/publisher.rb (hunk 1 above)
Comment (jjaffeux) [reply plugins/chat/app/services/chat/publisher.rb:116]: I think you are right that we alrzady have this info on the chat message now

File: plugins/chat/lib/chat_sdk/message.rb (hunk 7 above)
Comment (jjaffeux) [reply plugins/chat/lib/chat_sdk/message.rb:63]: Yes I hesitated, I think that's ok for now. But I wîll probably rethink this.

File: plugins/chat/assets/javascripts/discourse/components/chat-message.gjs (hunk 4 above)
Comment (jjaffeux) [reply plugins/chat/assets/javascripts/discourse/components/chat-message.gjs:492]: I think that potentially we could stream a message as a reply to another message if that's what you were asking?

**Review by dbattersby — APPROVED** [review]
Looking really good to me, I've run a few tests with streaming messages locally and it works great, nice work 👍

**Pull request merged**